File: PRN231-FP/FP-FAP/Models/Feedback.cs
namespace FP_FAP.Models;

using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;

[Table("Feedback")]
public class Feedback
{
    [Key]
    [DatabaseGenerated(DatabaseGeneratedOption.Identity)]
    public int Id { get; set; }

    [Required]
    [ForeignKey(nameof(User))]
    public int StudentId { get; set; }

    [Required]
    [ForeignKey(nameof(Models.Group))]
    public int GroupId { get; set; }

    public bool IsDone { get; set; }

    public int Punctuality { get; set; }

    public int Skill { get; set; }

    public int Adequately { get; set; }

    public int Support { get; set; }

    public int Response { get; set; }

    [MaxLength(500)]
    public string? Comment { get; set; }

    public virtual User Student { get; set; } = null!;

    public virtual Group Group { get; set; } = null!;
}

File: PRN231-FP/fp-user-client/src/api/fap-api/feedback.ts
import { baseURL } from ".";

export async function assignFeedback(accessToken: string, groupId: number, students: Array<number>): Promise<boolean> {
    try {
        const response = await fetch(
            `${baseURL()}/feedback/assign`,
            {
                method: "POST",
                headers: {
                    "Content-Type": "application/json",
                    Authorization: `Bearer ${accessToken}`,
                },
                body: JSON.stringify({
                    groupId,
                    students,
                }),
            }).then((res) => res.ok);
        return response;
    } catch (error: any) {
        return false;
    }
}

export async function doFeedback(accessToken: string, feedback: {
    id: number,
    studentId: number,
    groupId: number,
    isDone: boolean,
    punctuality: number,
    skills: number,
    adequately: number,
    support: number,
    response: number,
    comment: string,
}): Promise<boolean> {
    try {
        const response = await fetch(
            `${baseURL()}/feedback`,
            {
                method: "POST",
                headers: {
                    "Content-Type": "application/json",
                    Authorization: `Bearer ${accessToken}`,
                },
                body: JSON.stringify(feedback),
            }).then((res) => res.ok);
        return response;
    } catch (error: any) {
        return false;
    }
}

export async function getFeedback(accessToken: string, id: number): Promise<{
    id: number,
    studentId: number,
    groupId: number,
    isDone: boolean,
    punctuality: number,
    skills: number,
    adequately: number,
    support: number,
    response: number,
    comment: string,
    group: Array<{
        id: number,
        name: string,
        subjectId: number,
        teacherId: number,
        semester: string,
        subject: {
            id: number,
            name: string,
            code: string,
        },
        teacher: {
            id: number,
            email: string,
            role: "admin" | "teacher" | "student",
        },
    }>
} | null> {
    try {
        const response = await fetch(
            `${baseURL()}/feedback/${id}`,
            {
                method: "GET",
                headers: {
                    "Content-Type": "application/json",
                    Authorization: `Bearer ${accessToken}`,
                },
            }).then((res) => res.json());
        return response;
    } catch (error: any) {
        return null;
    }
}